source/unix/openssl_platform_init.c

```diff
@@ -9,43 +9,119 @@
 
 #include <aws/cal/private/opensslcrypto_common.h>
 
-static struct openssl_hmac_ctx_1_0_table hmac_ctx_1_0_table;
-static struct openssl_hmac_ctx_1_1_table hmac_ctx_1_1_table;
+static struct openssl_hmac_ctx_table hmac_ctx_table;
 static struct openssl_evp_md_ctx_table evp_md_ctx_table;
 
-struct openssl_hmac_ctx_1_0_table *g_aws_openssl_hmac_ctx_1_0_table = NULL;
-struct openssl_hmac_ctx_1_1_table *g_aws_openssl_hmac_ctx_1_1_table = NULL;
+struct openssl_hmac_ctx_table *g_aws_openssl_hmac_ctx_table = NULL;
 struct openssl_evp_md_ctx_table *g_aws_openssl_evp_md_ctx_table = NULL;
 
+/* libcrypto 1.1 stub for init */
+static void s_hmac_ctx_init_noop(HMAC_CTX *ctx) {
+    (void)ctx;
+}
+
+/* libcrypto 1.1 stub for clean_up */
+static void s_hmac_ctx_clean_up_noop(HMAC_CTX *ctx) {
+    (void)ctx;
+}
+
+/* libcrypto 1.0 shim for new */
+static HMAC_CTX *s_hmac_ctx_new(void) {
+    AWS_PRECONDITION(
+        g_aws_openssl_hmac_ctx_table->init_fn != s_hmac_ctx_init_noop &&
+        "libcrypto 1.0 init called on libcrypto 1.1 vtable");
+    HMAC_CTX *ctx = aws_mem_calloc(aws_default_allocator(), 1, 300);
+    AWS_FATAL_ASSERT(ctx && "Unable to allocate to HMAC_CTX");
+    g_aws_openssl_hmac_ctx_table->init_fn(ctx);
+    return ctx;
+}
+
+/* libcrypto 1.0 shim for free */
+static void s_hmac_ctx_free(HMAC_CTX *ctx) {
+    AWS_PRECONDITION(ctx);
+    AWS_PRECONDITION(
+        g_aws_openssl_hmac_ctx_table->clean_up_fn != s_hmac_ctx_clean_up_noop &&
+        "libcrypto 1.0 clean_up called on libcrypto 1.1 vtable");
+    g_aws_openssl_hmac_ctx_table->clean_up_fn(ctx);
+    aws_mem_release(aws_default_allocator(), ctx);
+}
+
+/* libcrypto 1.0 shim for reset */
+static void s_hmac_ctx_reset(HMAC_CTX *ctx) {
+    AWS_PRECONDITION(ctx);
+    AWS_PRECONDITION(
+        g_aws_openssl_hmac_ctx_table->init_fn != s_hmac_ctx_init_noop &&
+        g_aws_openssl_hmac_ctx_table->clean_up_fn != s_hmac_ctx_clean_up_noop &&
+        "libcrypto 1.0 reset called on libcrypto 1.1 vtable");
+    g_aws_openssl_hmac_ctx_table->clean_up_fn(ctx);
+    g_aws_openssl_hmac_ctx_table->init_fn(ctx);
+}
+
+void *s_find_libcrypto_module(void) {
+#if defined(AWS_CAL_EXPORTS)
```

Comment:
Isn't this guaranteed always true or false since this is a .c file and so it's only the internal view?

Reply:
It is only defined if we are building via `BUILD_SHARED=ON`